demos/signature/EVP_DSA_Signature_demo.c

```diff
@@ -0,0 +1,236 @@
+    if (pkey == NULL)
+        goto end;
+
+    result = 1;
+end:
+    if(!result) {
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+    }
+    EVP_PKEY_CTX_free(ctx);
+    *p_pkey = pkey;
+    fprintf(stdout, "Generating public/private key pair:\n");
+    EVP_PKEY_print_public_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+    EVP_PKEY_print_private_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+    EVP_PKEY_print_params_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+
+    return result;
+}
+
+static int demo_sign(OSSL_LIB_CTX *libctx,
+                     size_t *p_sig_len, unsigned char **p_sig_value,
+                     EVP_PKEY *pkey)
+{
+    int result = 0;
+    size_t sig_len = 0;
+    unsigned char *sig_value = NULL;
+    EVP_MD_CTX *ctx = NULL;
+
+    ctx = EVP_MD_CTX_create();
+    if (ctx == NULL)
+        goto end;
+
+    if (EVP_DigestSignInit_ex(ctx, NULL, DIGEST, libctx, NULL, pkey, NULL) != 1)
+        goto end;
+
+   if (EVP_DigestSignUpdate(ctx, hamlet_1, sizeof(hamlet_1)) != 1)
+        goto end;
+
+   if (EVP_DigestSignUpdate(ctx, hamlet_2, sizeof(hamlet_2)) != 1)
+        goto end;
+
+    if (EVP_DigestSignFinal(ctx, NULL, &sig_len) != 1)
+        goto end;
+    if (sig_len <= 0)
+        goto end;
+
+    sig_value = OPENSSL_malloc(sig_len);
+    if (sig_value == NULL)
+        goto end;
+
+    if (EVP_DigestSignFinal(ctx, sig_value, &sig_len) != 1)
+        goto end;
+
+    result = 1;
+end:
+    EVP_MD_CTX_destroy(ctx);
+    if (!result) {
+        OPENSSL_free(sig_value);
+        sig_len = 0;
+        sig_value = NULL;
+    }
+    *p_sig_len = sig_len;
+    *p_sig_value = sig_value;
+    fprintf(stdout, "Generating signature:\n");
+    BIO_dump_indent_fp(stdout, sig_value, sig_len, 2);
+    fprintf(stdout, "\n");
+    return result;
+}
+
+static int demo_verify(OSSL_LIB_CTX *libctx,
+                       size_t sig_len, unsigned char *sig_value,
+                       EVP_PKEY *pkey)
+{
+    int result = 0;
+    EVP_MD_CTX *ctx = NULL;
+
+    ctx = EVP_MD_CTX_create();
+    if(ctx == NULL)
+        goto end;
+
+    if (EVP_DigestVerifyInit_ex(ctx, NULL, DIGEST, libctx, NULL, pkey, NULL) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyUpdate(ctx, hamlet_1, sizeof(hamlet_1)) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyUpdate(ctx, hamlet_2, sizeof(hamlet_2)) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyFinal(ctx, sig_value, sig_len) != 1)
+        goto end;
+
+    result = 1;
+end:
+    EVP_MD_CTX_destroy(ctx);
+    return result;
+}
+
+int main(void)
+{
+    int result = 0;
+    OSSL_LIB_CTX *libctx = NULL;
+    EVP_PKEY *params = NULL;
+    EVP_PKEY *pkey = NULL;
+    size_t sig_len = 0;
+    unsigned char *sig_value = NULL;
+
+    libctx = OSSL_LIB_CTX_new();
+
+    if (!generate_dsa_params(libctx, &params))
+        goto end;
+
+    if (!generate_dsa_key(libctx, params, &pkey))
+        goto end;
+
+    if (!demo_sign(libctx, &sig_len, &sig_value, pkey))
```

Comment:
You might want to consider demonstrating how you would really use this..
i.e verify with a public key.. And put some comments such as

The signer signs with his private key, and distributes his public key..

A verifier uses the signers public key to verify the signature.

Reply:
(There are multiple ways to get the public key, EVP_PKEY_todata/fromdata is one way (evp_extra_test2.c does this) - or you could save and load).